DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/zero_copy_data_transfer/src/restricted_usm_kernel.hpp

```diff
@@ -0,0 +1,142 @@
+#ifndef __RESTRICTED_USM_KERNEL_HPP__
+#define __RESTRICTED_USM_KERNEL_HPP__
+#pragma once
+
+#include <vector>
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+
+using namespace sycl;
+using namespace std::chrono;
+
+//
+// The structure of the kernels in this design is shown in the diagram below.
+// The Producer kernel reads the data from CPU memory (via PCIe), producing it
+// for the RestrictedUSM via a pipe. The Worker does the computation on the
+// input data and writes it to the ConsumePipe. The consumer reads the data
+// from this pipe and writes the output back to the CPU memory (via PCIe).
+//
+//                                |----------------------------------|
+//                                |              FPGA                |
+//               |-------------|  |                                  |
+//               |             |  | |- --------|   |---------------| |
+//  |-------|    |             |--->| Producer |==>|               | |
+//  |       |    |             |  | |----------|   |               | |
+//  |  CPU  |<-->| Host Memory |  |                | RestrictedUSM | |
+//  |       |    |             |  | |----------|   |               | |
+//  |-------|    |             |<---| Consumer |<==|               | |
+//               |             |  | |----------|   |---------------| |
+//               |-------------|  |                                  |
+//                                |----------------------------------|
+//
+//
+// As shown in the image above and the code below, we have split this design
+// into three kernels:
+//    1) Producer
+//    2) RestrictedUSM
+//    3) Consumer
+// We do this to decouple the reads/writes from/to the Host Memory over PCIe.
+// Decoupling the memory accesses and using SYCL pipes with a substantial
+// depth ('kPipeDepth' below) allows the kernel to be more resilient against
+// stalls while reading/writing from/to the Host Memory over PCIe. 
+//
+
+// Declare the kernel class names globally to avoid name mangling.
+class RestrictedUSM;
+class Producer;
+class Consumer;
+
+// pipes
+constexpr size_t kPipeDepth = 64;
+template <typename T>
+using ProducePipe = pipe<class ProducePipeClass, T, kPipeDepth>;
+template <typename T>
+using ConsumePipe = pipe<class ConsumePipeClass, T, kPipeDepth>;
+
+//
+// reads the input data from the hosts memory
+// and writes it to the ProducePipe
+//
+template <typename T>
+event SubmitProducer(queue& q, T* in_data, size_t size) {
+  auto e = q.submit([&](handler& h) {
+    h.single_task<Producer>([=]() [[intel::kernel_args_restrict]] {
+      // using a host_ptr tells the compiler that this pointer lives in the
+      // hosts address space
+      host_ptr<T> h_in_data(in_data);
+
+      for (size_t i = 0; i < size; i++) {
+        const T data_from_host_memory = *(h_in_data + i);
```

Comment:
Why not h_in_data[i]  ?

Reply:
host_ptr operator[] has issues.

Reply:
I realize this was not very descriptive. Operator[] returns a const value, so using it gives this error for writing:

`error: cannot assign to return value because function 'operator[]' returns a const value h_out_data[i] = data_to_host_memory;`